File: go-version/leetcode/LC904/total_fruit.go
package LC904

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/8/4 下午11:22
 * @FilePath: total_fruit
 * @Description: 滑动窗口
 */

func totalFruit(fruits []int) int {
	var left, ans int
	mp := make(map[int]int)
	for i := 0; i < len(fruits); i++ {
		mp[fruits[i]]++
		for len(mp) > 2 {
			mp[fruits[left]]--
			if mp[fruits[left]] == 0 {
				delete(mp, fruits[left])
			}
			left++
		}
		ans = max(ans, i-left+1)
	}
	return ans
}

File: python-version/LC904/total_fruit.py
from collections import defaultdict
from typing import List


class Solution:
    @staticmethod
    def totalFruit(fruits: List[int]) -> int:
        left, ans = 0, 0
        mp = defaultdict(int)
        for i, v in enumerate(fruits):
            mp[v] = mp.get(v, 0) + 1
            while len(mp) > 2:
                mp[fruits[left]] = mp[fruits[left]] - 1
                if mp[fruits[left]] == 0:
                    del mp[fruits[left]]
                left += 1
            ans = max(ans, i - left + 1)
        return ans
